<div class="alert alert-block alert-info"> Reading in a short story as text sample into Python. </div>



In [1]:
with open("text_data/the-verdict.txt","r",encoding="utf-8") as f:
  raw_text = f.read()



## Step 1: Creating Tokens

<div class="alert alert-block alert-warning">
The print command prints the total number of characters followed by the first 100 characters of this file for illustration purposes.
</div>

In [2]:
print("total number of charater : ", len(raw_text))
print(raw_text[:99])

total number of charater :  20479
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no 


<div class="alert alert-block alert-success">So the goal is to tokenize this 20,479 character short story into individual words and special charaters  taht we an then turn inot embeddings for LLM training </div>

<div class="alert alert-block alert-info"> Using some simple example text, we can use the re.split command with the following syntax to split a text on whitespace </div>

In [3]:
import re

text = "Hello, world. This, is a test."
result = re.split(r'(\s)',text)
print(result)

['Hello,', ' ', 'world.', ' ', 'This,', ' ', 'is', ' ', 'a', ' ', 'test.']


<div class="alert alert-block alert-warning">The result is a list of individual words , whitespaces, and punctuation charaters </div>

In [4]:
# let's modify the regular expression splits on whitespaces (\s) and commans, and full stop
result = re.split(r'([,.]|\s)',text)
print(result)

['Hello', ',', '', ' ', 'world', '.', '', ' ', 'This', ',', '', ' ', 'is', ' ', 'a', ' ', 'test', '.', '']


we can see that the words and punctuation charaters are now separate list entry

<div class="alert alert-block alert-danger"> A small remaining  issue is that the list still includes whitespace characters. Optionally, we can remove these redundant characters safely as follows </div>

In [5]:
result = [item for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'This', ',', 'is', 'a', 'test', '.']


#### Removing whitespaces or not
<div class="alert alert-block alert-info">
When developing a simple tokenizer, whether we should encode whitespaces as separate characters or just remove them depends on our application and its requirements. Removing whitespaces reduces the memory and computing requirements. However, keeping whitespaces can be useful if we train models that are sensitive to the exact structure of the text (for example, Python code, which is sensitive to indentation and spacing). Here, we remove whitespaces for simplicity and brevity of the tokenized outputs. Later, we will switch to a tokenization scheme that includes whitespaces.

</div>

In [6]:
text = "Hello, world. Is this-- a test?"
result = re.split(r'([,.:;?_!"()\']|--|\s)',text)
result = [item.strip() for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?']


In [7]:
# Strip whitespace from each item and then filter out any empty strings.
result = [item for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?']


<div class="alert alert-block alert-warning">Testing with a sample text </div>

In [8]:
text = "Hello, world. Is this-- a test?"

In [9]:
result = re.split(r'([,.:;?_!"()\']|--|\s)', text)
result = [item.strip() for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?']


<div class="alert alert-block alert-success"> Now that we got a basic tokenizer working, let's apply it to Edith Wharton's entire short story: </div>

In [10]:
preprocessed = re.split(r'([,.:;?_\'"()\[\]|\s])', raw_text)
preprocessed = [item.strip() for item in preprocessed if item.strip()]
print(preprocessed[:30])

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius--though', 'a', 'good', 'fellow', 'enough--so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in', 'the', 'height', 'of', 'his']


In [11]:
print(len(preprocessed))

4481


# Step 2 : Creating Token IDs

<div class="alert alert-block alert-success">
In the previous section, we tokenized Edith Wharton's short story and assigned it to a Python variable called preprocessed.

Let's create a list of unique tokens and sort them alphabetically to determine the vocabulary size </div>

In [12]:
all_words = sorted(set(preprocessed))
vocab_size = len(all_words)
print(vocab_size)

1217


After determining that the vocabulary size is 1,217 via the above code, we create the vocabulary and print its 51 entries for illustration purpose.

In [13]:
vocab = {token:integer for integer, token in enumerate(all_words)}

In [14]:
for i,item in enumerate(vocab.items()):
  print(item)
  if i >=50:
    break

('"', 0)
("'", 1)
('(', 2)
(')', 3)
(',', 4)
('--and', 5)
('--it', 6)
('--oh', 7)
('--she', 8)
('--that', 9)
('.', 10)
(':', 11)
(';', 12)
('?', 13)
('A', 14)
('Ah', 15)
('Ah--I', 16)
('Among', 17)
('And', 18)
('Are', 19)
('Arrt', 20)
('As', 21)
('At', 22)
('Be', 23)
('Begin', 24)
('Burlington', 25)
('But', 26)
('By', 27)
('Carlo', 28)
('Chicago', 29)
('Claude', 30)
('Come', 31)
('Croft', 32)
('Destroyed', 33)
('Devonshire', 34)
('Don', 35)
('Dubarry', 36)
('Emperors', 37)
('Florence', 38)
('For', 39)
('Gallery', 40)
('Gideon', 41)
('Gisburn', 42)
('Gisburn!', 43)
('Gisburn--as', 44)
('Gisburn--fond', 45)
('Gisburns', 46)
('Grafton', 47)
('Greek', 48)
('Grindle', 49)
('Grindles', 50)


<div class="alert alert-block alert-success"> As we can see, based on the output above, the dictionary contains individual tokens associated with unique integer labels. </div>


<div class="alert alert-block alert-success"> Later, when we want to convert the outputs of an LLM from numbers back into text, we also need a way to turn token IDs back into text.

For this, we can create an inverse version of the vocabulary that maps token IDs back to corresponding text tokens. </div>

<div class="alert alert-block alert-success"> Let's implement a complete tokenizer class in Python.

The class will have an encode method that splits text into tokens and carries out the string-to-integer mapping to produce the tokens

In addition, we implement a decode method that carries out the reverse integer-to-string mapping to convert the token Id back to the string </div>

In [15]:
class SimpleTokenizerV1:
  def __init__(self,vocab):
    self.str_to_int = vocab
    self.int_to_str = {i:s for s,i in vocab.items()}


  def encode(self,text):
    preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)',text)

    preprocessed = [
        item.strip() for item in preprocessed if item.strip()
    ]
    ids = [self.str_to_int[s] for s in preprocessed]
    return ids


  def decode(self,ids):
    text = " ".join([self.int_to_str[i] for i in ids])
    # Replace spaces before the specified punctuations

    text = re.sub(r'\s+([,.?!"()\'])',r'\1',text)
    return text


Let's instantiate a new tokenizer object from the SimpleTokenizerV1 class and tokenizer a passage from Edith Wharton short story to try it out in practice

In [16]:
tokenizer = SimpleTokenizerV1(vocab)

text = """
"It's the last he painted,you know,"
Mrs. Gisburn said with pardonable pride."""



In [17]:
ids = tokenizer.encode(text)
print(ids)

[0, 63, 1, 920, 1065, 646, 563, 808, 4, 1212, 640, 4, 0, 78, 10, 42, 921, 1191, 818, 863, 10]


<div class="alert alert-block alert-success"> The code above prints the following token IDs: Next, let's see if we can turn these token IDs back into text using the decode method: </div>

In [18]:
tokenizer.decode(ids)

'" It\' s the last he painted, you know," Mrs. Gisburn said with pardonable pride.'

<div class="alert alert-block alert-success"> Based on the output above, we can see that the decode method successfully converted the token IDs back into the original text. </div>

<div class="alert alert-block alert-success"> So far, so good. We implemented a tokenizer capable of tokenizing and de-tokenizing text based on a snippet from the training set: </div>

In [19]:
# text = "Hello, do you like phone?"
# print(tokenizer.encode(text))

<div class="alert alert-block alert-danger"> The problem is that the world <b> "Hello" </b> was not used in the <b>The Verdict </b> short story.

Hence, it is not contained in the vocabulary.

This highlights the need to consider large and diverse training sets to extend the vocabulary when working on LLMs. </div>

## ADDING SPECIAL CONTEXT TOKENS
<div class="alert alert-block alert-info">
In the previous section, I have Implemented a simple tokenizer and applied it to a passage from the training set.

In this section, we will modify this tokenizer to handle unknown words.

In particular, we will modify the vocabulary and tokenizer we implemented in the previous section, SimpleTokenizerV2, to support two new tokens, <b><|unk|></b> and <b><|endoftext|></b> </div>




we can modify the tokenizer to use an <|unk|> token if it encounters a word that is not part of the vocabulary.

Furthermore, we add a token between unrealated texts.

For example, when training GPT-like LLMs on multiple independent documents or books , it is common to insert before each document or book that follows a previous text source

#### Let's now modify the vocabulary to include these two special tokens, and <|endoftext|>, by adding the previous section

In [20]:
all_tokens = sorted(list(set(preprocessed)))
all_tokens.extend(["<|endiftext|>","<|unk|>"])

vocab = {token:integer for integer,token in enumerate(all_tokens)}

In [21]:
len(vocab.items())

1219

<div class="alert alert-block alert-success"> Based on the output of the print statement above, the new vocabulary size is 1219 </div>

<div class="alert alert-block alert-warning"> As an additional quick check, let's print the last 5 entries of the updated vocabulary </div>

In [22]:
for i, item in enumerate(list(vocab.items())[-5:]):
  print(item)

('younger', 1214)
('your', 1215)
('yourself', 1216)
('<|endiftext|>', 1217)
('<|unk|>', 1218)


#### A simple text tokenizer than handles unknown words

Step 1: Replace unknown words by <|unk|> tokens

Step 2: Replace spaces before the specified punctuations

In [23]:
class SimpleTokenizerV2:
    def __init__(self, vocab):
        self.str_to_int = vocab
        self.int_to_str = {i: s for s, i in vocab.items()}

    def encode(self, text):
        preprocessed = re.split(r'([,.:;?_\'"()\[\]|\s])', text)
        preprocessed = [item.strip() for item in preprocessed if item.strip()]
        preprocessed = [
            item if item in self.str_to_int else "<|unk|>" for item in preprocessed
        ]
        ids = [self.str_to_int[s] for s in preprocessed]
        return ids
    def decode(self,ids):
      text = " ".join([self.int_to_str[i] for i in ids])
      # Replace spaces before the specified punctuations

      text = re.sub(r'\s+([,.?!"()\'])',r'\1',text)
      return text

In [24]:
tokenizer = SimpleTokenizerV2(vocab)

text1 = "Hello, do you like tea?"
text2 = "In the sunlit terraces of the palace."

text = "<|endoftext|>".join((text1,text2))

In [25]:
print(text)

Hello, do you like tea?<|endoftext|>In the sunlit terraces of the palace.


In [26]:
tokenizer.encode(text)

[1218,
 4,
 381,
 1212,
 673,
 1052,
 13,
 1218,
 1218,
 1218,
 1218,
 1218,
 1065,
 1031,
 1061,
 778,
 1065,
 1218,
 10]

In [27]:
tokenizer.decode(tokenizer.encode(text))

'<|unk|>, do you like tea? <|unk|> <|unk|> <|unk|> <|unk|> <|unk|> the sunlit terraces of the <|unk|>.'

<div class="alert alert-block alert-info">Based on comparing the de-tokenized text above with the original input text, we know that the training dataset, Edith Wharton's short story The Verdict, did not contain the words <b> "Hello"</b> and <b> "palace"</b> </div>

So far, we have discussed tokenization as an essential step in processing text as input to LLMs. Depending on the LLM, some researchers also consider additional special tokens such as the following:

- **[BOS] (beginning of sequence):** This token marks the start of a text. It signifies to the LLM where a piece of content begins.

- **[EOS] (end of sequence):** This token is positioned at the end of a text and is especially useful when concatenating multiple unrelated texts, similar to `<endoftext>`. For instance, when combining two different Wikipedia articles or books, the `[EOS]` token indicates where one article ends and the next one begins.

- **[PAD] (padding):** When training LLMs with batch sizes larger than one, the batch might contain texts of varying lengths. To ensure all texts have the same length, the shorter texts are extended or "padded" using the `[PAD]` token, up to the length of the longest text in the batch.


<div class="alert alert-block alert-info"> Note that the tokenizer used for GPT models does not need any of these tokens mentioned above but only uses an <b> <|endoftext|> </b> token for simplicity </div>

<div class="alert alert-block alert-info"> The tokenizer used for GPT models also doesn't use an <|unk|> token for out ofvocabulary words. Instead, GPT model use a <b> byte pair encoding tokenizer </b>, which breaks down words into subword units </div>

# BYTE PAIR ENCODING

<div class="alert alert-block alert-success"> 
We have implemented a <b> Simple Tokenization </b> scheme in the previous sections for illustraton purpose.

This section covers a more sophisticated tokenization scheme based on a concept called <b>byte pair encoding (BPE).</b>

The BPE tokenizer covered in this section was used to train LLMs such as GPT-2,GPT-3, and the original model used in ChatGPT

</div>

<div class="alert alert-block alert-warning"> 
So The implementing BPE can be relatively complicated, we will use an existing Python open-source library called tiktoken (openai uses it )

This library implements the BPE algorithm very efficiently basaed on source code in Rust.
</div>

In [28]:
!pip3 install tiktoken

'DOSKEY' is not recognized as an internal or external command,
operable program or batch file.


In [29]:
import importlib
import tiktoken

In [30]:
print("tiktoken version:",importlib.metadata.version("tiktoken"))

tiktoken version: 0.7.0


<div class="alert alert-block alert-success"> 
Once installed , we can instantiate the BPE tokenizer from tiktoken as follows:

</div>

In [31]:
tokenizer = tiktoken.get_encoding("gpt2")

<div class="alert alert-block alert-success"> 
The usage of this tokenizer is similar to SimpleTokenizerV2 we implemented previously via an encoder method:
</div>

In [32]:
text = (
    "Hello, do you like tea? <|endoftext|> In the sunlit terraces"
    "of some unknown Place."
)

integers = tokenizer.encode(text,allowed_special={"<|endoftext|>"})
print(integers)

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 1659, 617, 6439, 8474, 13]


<div class="alert alert-block alert-info"> 
The code above prints the following token IDs:
</div>

<div class="alert alert-block alert-success"> 
We can then convert the token IDs back into text using the decode method, similar to our SimpleTokenizerV2
</div>

In [33]:
strings = tokenizer.decode(integers)
print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terracesof some unknown Place.


<div class="alert alert-block alert-warning"> 
We can make  two note worthy observations based on the token IDs and decoded text above.

First, the <|endoftext> token is assigned a relatively large token ID, namely , 50256.

In fact, the BPE tokenizer, which was used to train models such as GPT-2, GPT-3, and the original model used in ChatGPT, hasa a total vocabulary size of 50,257

with <|endoftext> being assigned the largest token ID.

</div>

<div class="alert alert-block alert-warning"> 
Second, the BPE tokenizer above encodes and decodes  unknown words, such as "someunknownPlace" correctly.

The BPE tokenizer can handle any unknown word. How does it achieve this without using <|unk|> tokens ?

</div>

 <div class="alert alert-block alert-warning"> 
The algorithm underlying BPE break down words that aren't in its predefined vocabulary into smaller subwords units or even individual characters

This enables it to handle out-of-vocabulary words.

So, thanks to the BPE algorithm, if the tokenizer encounters an unfamiliear word during tokenization, it can represent it as a sequence of subword tokens or characters
</div>

**Let us take another simple example to illustrate how the BPE tokenizer deals with unknown tokens**

In [34]:
integers = tokenizer.encode("werva esd")
print(integers)



strings = tokenizer.decode(integers)
print(strings)

[86, 32775, 1658, 67]
werva esd


## CREATING INPUT-TARGET PAIRS

<div class="alert alert-block alert-success"> 
In this section we implement a data loader that fetches the input-target pairs using a sliding window approach.
</div>

<div class="alert alert-block alert-success"> 
To get started, we will first tokenize the whole <b> The Verdict </b> short story we worked with earlier using the BPE tokenizer
</div>

In [35]:
with open("text_data/the-verdict.txt","r",encoding="utf-8") as f:
    raw_text = f.read()

enc_text = tokenizer.encode(raw_text)
print(len(enc_text))


5145


<div class="alert alert-block alert-info"> 
Executing the code above  will return 5145, the total number of tokens in the training set, applying the BPE tokenizer.
</div>

<div class="alert alert-block alert-success"> 
Next, we remove the first 50 tokens from the dataset for demonstration purposes as it results in a slightly more interesting text passage in the next step
</div>

In [36]:
enc_sample = enc_text[50:]

<div class="alert alert-block alert-success"> 
One of the easiest and most intuitive ways to create the input-target pairs for the next word prediction task is to create two variables x and y where x contains the input tokens and y contains the target, which are the inputs shifted by 1:
</div>

<div class="alert alert-block alert-info"> 
The context size determines how many tokens are included in the input
</div>

In [37]:
context_size = 4 # length of the input

#the context_size of 4 means that the model is trained to look at a sequence of 4 words (or tokens)
#to predict the next word in the sequence.
#The input x is the first 4 toekns [1,2,3,4], and the target y is the next 4 tokens [2,3,4,5]

x = enc_sample[:context_size]
y = enc_sample[1:context_size+1]

In [38]:
print(f"x:{x}")
print(f"y:     {y}")

x:[290, 4920, 2241, 287]
y:     [4920, 2241, 287, 257]


<div class="alert alert-block alert-success"> 
Processing the inputs along with the targets, which are the inputs shifted by one position, we can then create the next-word prediction tasks as follows:
</div>

In [39]:
for i in range(1,context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]

    print(context,"---->", desired)


[290] ----> 4920
[290, 4920] ----> 2241
[290, 4920, 2241] ----> 287
[290, 4920, 2241, 287] ----> 257


<div class="alert alert-block alert-info"> 
Everything left of the arrow (--->) refers to the input an LLM would receive, and the token ID on the right side of the arrow represents the target token ID that the LLM is supposed to predict.

</div>

<div class="alert alert-block alert-success"> 
For illustration purposes, let's repeat the previous code but convert the token IDs into text:


</div>

In [40]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]

    print(tokenizer.decode(context),"---->", tokenizer.decode([desired]))
    

 and ---->  established
 and established ---->  himself
 and established himself ---->  in
 and established himself in ---->  a


<div class="alert alert-block alert-warning"> 
We've now created the input-target pairs that we can turn into use for the LLM training in upcoming chapters.


</div>

<div class="alert alert-block alert-warning"> 
There's only one more task before we can turn the tokens into embeddings: implementing an efficient data loader that iterates over the input dataset and returns the inputs and targets as Pyorch tensors, which can be throught of as multidimensional arrays.

</div>

<div class="alert alert-block alert-warning"> 
In particular, we are intereseted in returning two tensors: an input tensor containing thet text the LLM sees and a target tensor that includes the targets for the LLm to predict


</div>

## IMPLEMENTING A DATA LOADER

<div class="alert alert-block alert-success"> 
For the efficient data loader implementation, we will use PyTorch's built-in Dataset and DataLoader classes.


</div>

<div class="alert alert-block alert-info"> 

Step 1: Tokenize the entire text

Step 2: Use a sliding window to chunk the book into overlapping sequences of max_length

Step 3: Return the total number of rows in the dataset

Step 4:  Return a single row from the datset

</div>

In [41]:
from torch.utils.data import Dataset, DataLoader

class GPTDatasetV1(Dataset):
    def __init__(self,txt,tokenizer,max_length,stride):
        self.input_ids = []
        self.target_ids = []

        # Tokenize the entire text
        token_ids = tokenizer.encode(txt,allowed_special={"<|endoftext|>"})


        # use a sliding window to chunk the book into overlapping sequences of max_length
        for i in range(0,len(token_ids) - max_length, stride):
            input_chunk = token_ids[i:i + max_length]
            target_chunk = token_ids[i + 1: i + max_length + 1]
            self.input_ids.append(torch.tensor(input_chunk))
            self.target_ids.append(torch.tensor(target_chunk))




    def __len__(self):
        return len(self.input_ids)
    
    def __getitem__(self,idx):
        return self.input_ids[idx], self.target_ids[idx]

<div class="alert alert-block alert-warning"> 

The GPT Dataset V1 class in listing 2.5 is based on the PyTorch Dataset class.

It defines how individual rows are fetched from the dataset.

Each row consists of a number of token IDs (based on a max_length) assigned to an input_chunk tensor.

The target_chunk tensor contains the corresponding targets.

I recommend reading on to see how the data returned from this dataset looks like when we combine the dataset with additional intuition and clarity.

</div>

<div class="alert alert-block alert-success"> 
The following code will use the GPT Dataset V1 to load the inputs in batches via a PyTorch DataLoader:
</div>

<div class="alert alert-block alert-info"> 

Step 1: Initialize the tokenizer

Step 2: Create dataset

Step 3: drop_last=True drops the last batch if it is shorter than the specified batch_size to prevent loss spikes during training 

Step 4:  The number of CPU processes to use for preprocessing

</div>

In [42]:
def create_dataloader_v1(txt, batch_size=4,max_length=256,
                         stride=128, shuffle=True,drop_last=True, num_workers=0
                         ):
    # Initialize the tokenizer
    tokenizer = tiktoken.get_encoding("gpt2")

    # Create dataset
    dataset = GPTDatasetV1(txt,tokenizer,max_length,stride)


    # Create dataloader

    dataloader = DataLoader(
        dataset,
        batch_size=batch_size,
        shuffle=shuffle,
        drop_last=drop_last,
        num_workers=num_workers
    )


    return dataloader

<div class="alert alert-block alert-success"> 

Let's test the dataloader with a batch size of 1 for the LLM with a context size of 4,

This will develop an intuition of how the GPTDatasetV1 class and the create_dataloader_v1 function work together:
</div>

In [43]:
with open("text_data/the-verdict.txt","r",encoding="utf-8") as f:
    raw_text = f.read()

<div class="alert alert-block alert-info"> 
Convert dataloader into a Python iterator to fetch the next entry via Python's build-in next() function
</div>

In [44]:
import torch
print("PyTorch version :",torch.__version__)

dataloader = create_dataloader_v1(
    raw_text,batch_size=1,max_length=4,stride=1,shuffle=False
)

data_iter = iter(dataloader)
first_batch = next(data_iter)
print(first_batch)

PyTorch version : 2.4.1+cpu
[tensor([[  40,  367, 2885, 1464]]), tensor([[ 367, 2885, 1464, 1807]])]


<div class="alert alert-block alert-warning"> 
The first_batch variable contains two tensors: the first tensor stores the input token IDs, and the second tensor stores the target token IDs.

Since the max_length is set to 4, each of the two tensors contains 4 token IDs.

Note that an input size of 4 is relatively small and only chosen for illustration purposes. It is common to train LLMs with input sizes of at least <b> 256 </b>

</div>

<div class="alert alert-block alert-success"> 
To illustrate the meaning of stride=1, 
Let's fetch another batch from this dataset:
</div>

In [45]:
second_batch = next(data_iter)
print(second_batch)

[tensor([[ 367, 2885, 1464, 1807]]), tensor([[2885, 1464, 1807, 3619]])]


<div class="alert alert-block alert-warning"> 
If we compare the first with the second batch, we can see that the second batch's token IDs are shifted by one position compared to the first batch.

For example, the second ID in the first batch's input is 367, which is the first ID of the second batch's input.

The stride setting dictates the number of positions the inputs shift across batches, emulating a sliding window approach

</div>

<div class="alert alert-block alert-warning"> 
Batch sizes of 1, such as we have sampled from the data loader so far, are useful for illustration purposes.

If you have previous experience with deep learning, you may know that small batch sizes require less memory during training but lead to more noise  model updates.

Just like in regular deep learning, the batch size is a trade-off and hyperparameter to experiment with when training LLMs.
</div>

<div class="alert alert-block alert-success"> 
Before we move on to the two final  sections of this chapter that are focused on creating the embeddings vectors from the token IDs, let's have a brief look at how we can use the data loader to sample with a batch size greater than 1:
</div>

In [46]:
dataloader = create_dataloader_v1(raw_text,batch_size=8,max_length=4,stride=4,shuffle=False)


data_iter = iter(dataloader)
inputs,target = next(data_iter)
print("Inputs:\n",inputs)
print("\nTargets:\n",target)

Inputs:
 tensor([[   40,   367,  2885,  1464],
        [ 1807,  3619,   402,   271],
        [10899,  2138,   257,  7026],
        [15632,   438,  2016,   257],
        [  922,  5891,  1576,   438],
        [  568,   340,   373,   645],
        [ 1049,  5975,   284,   502],
        [  284,  3285,   326,    11]])

Targets:
 tensor([[  367,  2885,  1464,  1807],
        [ 3619,   402,   271, 10899],
        [ 2138,   257,  7026, 15632],
        [  438,  2016,   257,   922],
        [ 5891,  1576,   438,   568],
        [  340,   373,   645,  1049],
        [ 5975,   284,   502,   284],
        [ 3285,   326,    11,   287]])


<div class="alert alert-block alert-warning"> 
Note that we increase the stride to 4. This is to utilize the dataset fully (we don't skip a single word) but also avoid any overlap between the batches, since more overlap could lead to increased overfitting.
</div>

## CREATING TOKEN EMBEDDINGS

<div class="alert alert-block alert-success"> 
Let's illustrate how the token ID to embedding  vector conversion works with a hands-on example.

Suppose we have the following four input tokens with IDs 2,3,5 and 1
</div>

In [47]:
input_ids = torch.tensor([2,3,5,1])

<div class="alert alert-block alert-success"> 
For the sake of simplicity and illustration purposes, suppose we have a small vocabulary of only 6 words (instead of the 50,257 words in the  vocabulary), and we want to create embeddings of size 3 (in GPT-3, the embedding size is 12,288 dimensins):
</div>

<div class="alert alert-block alert-success"> 
Using the vocab_size and output_dim, we can instantiate an embedding layer in PyTorch, setting the random seed to 123 for reproducibility purposes:


</div>

In [48]:
vocab_size = 6
output_dim = 3

torch.manual_seed(123)

embedding_layer = torch.nn.Embedding(vocab_size,output_dim)

<div class="alert alert-block alert-info"> 
The Print statement in the code prints the embeddings layer;s underlying weight matrix:
</div>

In [49]:
print(embedding_layer.weight)

Parameter containing:
tensor([[ 0.3374, -0.1778, -0.1690],
        [ 0.9178,  1.5810,  1.3010],
        [ 1.2753, -0.2010, -0.1606],
        [-0.4015,  0.9666, -1.1481],
        [-1.1589,  0.3255, -0.6315],
        [-2.8400, -0.7849, -1.4096]], requires_grad=True)


<div class="alert alert-block alert-info"> 
We can see that the weight matrix of the embedding layer contains small, random values. 
These values are optimized during LLM training as part of the LLM optimization itself, as we will see in upcomin chapters. 
Moreover, we can see that the weight matrix hs six rows and three columns. There is one row for each of
the six possilbe tokens in the vocabulary. And there is one column for each of the three embedding dimensions.
</div>

<div class="alert alert-block alert-success"> 
After we instantiated the embedding layer, let's now apply it to a token ID to optain the embedding vector:
</div>

In [50]:
print(embedding_layer(torch.tensor([3])))

tensor([[-0.4015,  0.9666, -1.1481]], grad_fn=<EmbeddingBackward0>)


<div class="alert alert-block alert-info"> 
if we compare the embedding vector for token ID 3 to the previous embedding matrix, we see that it is identical to the 4th row (Python starts with a zero)
it's the row corresponding to index 3. 
In other words, the embedding layer is essentially a look-up operation that matrix via a token ID.
</div>

In [51]:
print(embedding_layer(input_ids))

tensor([[ 1.2753, -0.2010, -0.1606],
        [-0.4015,  0.9666, -1.1481],
        [-2.8400, -0.7849, -1.4096],
        [ 0.9178,  1.5810,  1.3010]], grad_fn=<EmbeddingBackward0>)


<div class="alert alert-block alert-info"> 
Each row in this output matrix is obtained via a lookup operation from the embedding weight matrix
</div>